enterprise/internal/insights/background/retention/worker.go

```diff
@@ -7,36 +7,72 @@ import (
 	"github.com/keegancsmith/sqlf"
 	"github.com/sourcegraph/log"
 
+	"github.com/sourcegraph/sourcegraph/enterprise/internal/insights/store"
+	"github.com/sourcegraph/sourcegraph/internal/conf"
 	"github.com/sourcegraph/sourcegraph/internal/database/basestore"
 	"github.com/sourcegraph/sourcegraph/internal/observation"
 	"github.com/sourcegraph/sourcegraph/internal/workerutil"
 	"github.com/sourcegraph/sourcegraph/internal/workerutil/dbworker"
 	dbworkerstore "github.com/sourcegraph/sourcegraph/internal/workerutil/dbworker/store"
+	"github.com/sourcegraph/sourcegraph/lib/errors"
 )
 
 var _ workerutil.Handler[*DataRetentionJob] = &dataRetentionHandler{}
 
 type dataRetentionHandler struct {
 	baseWorkerStore dbworkerstore.Store[*DataRetentionJob]
+	insightsStore   *store.Store
 }
 
-func (h *dataRetentionHandler) Handle(ctx context.Context, logger log.Logger, record *DataRetentionJob) error {
-	logger.Debug("data retention handler called", log.Int("seriesID", record.SeriesID))
+func (h *dataRetentionHandler) Handle(ctx context.Context, logger log.Logger, record *DataRetentionJob) (err error) {
+	// Default should match what is shown in the schema not to be confusing
+	maximumSampleSize := 90
+	if configured := conf.Get().InsightsMaximumSampleSize; configured != 0 {
+		maximumSampleSize = configured
+	}
+
+	// All the retention operations need to be completed in the same transaction
+	tx, err := h.insightsStore.Transact(ctx)
+	if err != nil {
+		return err
+	}
+	defer func() { err = tx.Done(err) }()
+
+	oldestRecordingTime, err := selectOldestRecordingTimeBeforeMax(ctx, tx, record.InsightSeriesID, maximumSampleSize)
+	if err != nil {
+		return errors.Wrap(err, "selectOldestRecordingTimeBeforeMax")
+	}
+
+	if oldestRecordingTime == nil {
+		// this series does not have any data beyond the max sample size
+		logger.Debug("data retention procedure not needed", log.Int("seriesID", record.InsightSeriesID), log.Int("maxSampleSize", maximumSampleSize))
+		return nil
+	}
+
+	if err := archiveOldRecordingTimes(ctx, tx, record.InsightSeriesID, *oldestRecordingTime); err != nil {
+		return err
+	}
+
+	if err := archiveOldSeriesPoints(ctx, tx, record.SeriesID, *oldestRecordingTime); err != nil {
+		return err
+	}
+
 	return nil
 }
 
 // NewWorker returns a worker that will find what data to prune and separate for a series.
-func NewWorker(ctx context.Context, logger log.Logger, workerStore dbworkerstore.Store[*DataRetentionJob], metrics workerutil.WorkerObservability) *workerutil.Worker[*DataRetentionJob] {
+func NewWorker(ctx context.Context, logger log.Logger, workerStore dbworkerstore.Store[*DataRetentionJob], insightsStore *store.Store, metrics workerutil.WorkerObservability) *workerutil.Worker[*DataRetentionJob] {
 	options := workerutil.WorkerOptions{
 		Name:              "insights_data_retention_worker",
 		NumHandlers:       5,
-		Interval:          5 * time.Second,
+		Interval:          12 * time.Hour,
```

Comment:
I don't think this is often enough, the queue will back up if it only checks for work every 12 hours.

Reply:
mb I put in a somewhat arbitrary value late at the end of my PR
```suggestion
		Interval:          30 * time.Minute,
```